composer_migration/src/composer_migration/lib/strategies/migrate_kubernetes_pod_operator_ast.py

```diff
@@ -0,0 +1,122 @@
+class KPOFinder(ast.NodeVisitor):
+    def __init__(self: ast.NodeVisitor) -> None:
+        self.has_problem = False
+        self.kpo_attribute = False
+
+    # Binding to ast.AST means we can accept any node type
+    def visit_value(self: ast.NodeVisitor, node: A) -> None:
+        print(type(node))
+        self.generic_visit(node)
+
+    def visit_Call(self: ast.NodeVisitor, node: ast.Call) -> None:
+        self.generic_visit(node)
+
+    # Airflow 2 nodes will determine this way
+    def visit_Name(self: ast.NodeVisitor, node: ast.Name) -> None:
+        if node.id == "KubernetesPodOperator":
+            self.kpo_attribute = True
+            self.generic_visit(node)
+
+    # Airflow 1 nodes will determine this way
+    def visit_Attribute(self: ast.NodeVisitor, node: ast.Attribute) -> None:
+        if node.attr == "KubernetesPodOperator":
+            self.kpo_attribute = True
+            self.generic_visit(node)
+
+    def visit_keyword(self: ast.NodeVisitor, node: ast.keyword) -> None:
+        if node.arg == "affinity":
+            self.generic_visit(node)
+
+    def visit_Dict(self: ast.NodeVisitor, node: ast.Dict) -> None:
+        if len(node.keys) != 0 and self.kpo_attribute:  # This means pod affinity is set
+            self.has_problem = True
+            self.kpo_attribute = False
+
+
+class CheckKubernetesPodOperator(DAGChecker):
+    def check_for_problem(file: str) -> List[str]:
+        TABLE_TITLE = "KubernetesPodOperator Affinity Check"
+        TABLE_TEXT = "Node pools are not supported with the KubernetesPodOperator in Composer 2. Refactor the following DAGs to use the GKE Operators to launch your pod in a new cluster, or remove the 'affinity' field from your operator. See https://cloud.google.com/composer/docs/composer-2/use-gke-operator for more information on using the GKE operators in Cloud Composer."
+        # read dag in as a string
+        with open(file, "r") as f:
+            dag = f.read()
+        # parse string into abstract syntax tree
+        tree = ast.parse(dag)
+
+        # create a traverser that looks for "ast.Assign" nodes
+        a = AssignFinder()
+
+        # visit the ast and add all assign nodes to a list
+        # we need the "assign" nodes to be able to name the problem variables
+        a.visit(tree)
+
+        # go through possible problem notes and check for "affinity" parameter
+        problem_nodes = []
+        for assign in a.possible_kpo_nodes:
+            k = KPOFinder()
+            k.visit(assign)
+            # assumes only one target ever
+            # if target node doesn't have an ID, log a debug statement
+            try:
+                i = assign.targets[0].id
+                if k.has_problem:
+                    problem_nodes.append(i)
+            except AttributeError:
+                logging.debug("Node is not of type with a target")
+
+        output = {"nodes": problem_nodes, "title": TABLE_TITLE, "text": TABLE_TEXT}
```

Comment:
Does this match the method signature, return type = "List[str]" ? 

Looks like a dictionary is being return ?

Reply:
Ha, you are very right. It does not! Great catch.